cfgrammar/src/lib/newlinecache.rs

```diff
@@ -121,6 +122,24 @@ impl NewlineCache {
             (line_num, column)
         })
     }
+
+    /// Return the (start byte, end byte) of the lines containing `span`. This will always cover
+    /// at least 1 logical line.
+    pub fn span_line_bytes(&self, span: Span) -> (usize, usize) {
+        let (st, st_line) = match self.newlines.binary_search(&span.start()) {
+            Ok(j) => (self.newlines[j], j + 1),
+            Err(0) => (0, 0),
+            Err(j) => (self.newlines[j - 1], j),
+        };
+        let en = match self.newlines[st_line..].binary_search(&span.end()) {
+            Ok(j) => self.newlines[st_line + j + 1] - 1,
+            Err(j) if st_line + j == self.newlines.len() => {
+                self.newlines.last().unwrap() + self.trailing_bytes
+            }
+            Err(j) => self.newlines[st_line + j] - 1,
```

Comment:
I've had a little trouble convincing myself the `Ok(j)` and `Err(j)` cases where the slicing or subtraction could cause a panic are unreachable, and covered by the pattern guard.

But, given that this is movement from below, and the conditions of the `newlines` should be the same, 
It now also handles `trailing_bytes`, So I wouldn't consider any issues here to be a regression.

So, I think this patch as it is an improvement, but it might be nice to have some clarification around these aspects, but given that it is moving code perhaps that could be a future improvement

Reply:
It seems like `span_line_eq("aa\n", 2, 3, "aa\n");` hits one of the panics I was worried about in the `Ok(j)` branch.

Reply:
Very good spot. Fixed (with a mild difference, but the same in spirit) in 01c24eb.